src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs

```diff
@@ -572,12 +572,31 @@ private async Task<bool> EnsureConnectionStateAsync(HttpConnectionContext connec
 
         private static void CloneUser(HttpContext newContext, HttpContext oldContext)
         {
-            if (oldContext.User.Identity is WindowsIdentity)
+            // If the identity is a WindowsIdentity we need to clone the User.
+            // This is because the WindowsIdentity uses SafeHandle's which are disposed at the end of the request
+            // and accessing the identity can happen outside of the request scope.
+            if (oldContext.User.Identity is WindowsIdentity windowsIdentity)
             {
-                newContext.User = new ClaimsPrincipal();
+                var skipFirstIdentity = false;
+                if (oldContext.User is WindowsPrincipal)
+                {
+                    // We want to explicitly create a WindowsPrincipal instead of a ClaimsPrincipal
+                    // so methods that WindowsPrincipal overrides like 'IsInRole', work as expected.
+                    newContext.User = new WindowsPrincipal((WindowsIdentity)(windowsIdentity.Clone()));
+                    skipFirstIdentity = true;
+                }
+                else
+                {
+                    newContext.User = new ClaimsPrincipal();
```

Comment:
Nit, to get rid of skipFirstIdentity

```suggestion
                    newContext.User = new ClaimsPrincipal(windowsIdentity.Clone());
```

Reply:
Think I have to keep it as is, because I still need to skip somehow when copying the rest of the identities.

Reply:
Wouldn't you just always skip? You could do something like `oldContext.User.Identities.Skip(1)`, right?

Reply:
That allocates I believe